File: architecture/modules/storage/go/http.go
package storage

import (
	"encoding/json"
	"errors"
	"io"
	"mime"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// Authenticate must verify the project session/token and return its stable user ID.
type HTTPOptions struct {
	Service          *FileService
	Authenticate     func(*http.Request) (string, error)
	BasePath         string
	TrustedOrigins   []string
	AllowCredentials bool
}

func NewFileHandler(options HTTPOptions) (http.Handler, error) {
	base := options.BasePath
	if base == "" {
		base = "/files"
	}
	if !strings.HasPrefix(base, "/") || strings.HasSuffix(base, "/") || strings.ContainsAny(base, "?# ") || options.Service == nil || options.Authenticate == nil {
		return nil, fail("CONFIGURATION")
	}
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != base && !strings.HasPrefix(r.URL.Path, base+"/") {
			http.NotFound(w, r)
			return
		}
		requestID := uuid()
		w.Header().Set("X-Request-ID", requestID)
		w.Header().Set("Cache-Control", "no-store")
		w.Header().Set("X-Content-Type-Options", "nosniff")
		send := func(v any, status int) {
			if status == 204 {
				w.WriteHeader(status)
				return
			}
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(status)
			_ = json.NewEncoder(w).Encode(v)
		}
		operation := func() error {
			if origin := r.Header.Get("Origin"); origin != "" {
				allowed := false
				for _, o := range options.TrustedOrigins {
					if origin == o {
						allowed = true
					}
				}
				if !allowed {
					return fail("FORBIDDEN")
				}
				w.Header().Set("Access-Control-Allow-Origin", origin)
				w.Header().Set("Vary", "Origin")
				if options.AllowCredentials {
					w.Header().Set("Access-Control-Allow-Credentials", "true")
				}
			}
			if r.Method == "OPTIONS" {
				w.Header().Set("Access-Control-Allow-Methods", "GET,POST,PUT,DELETE,OPTIONS")
				w.Header().Set("Access-Control-Allow-Headers", "authorization,content-type")
				send(nil, 204)
				return nil
			}
			actor, e := options.Authenticate(r)
			if e != nil || actor == "" {
				return fail("FORBIDDEN")
			}
			ctx := r.Context()
			parts := strings.Split(strings.TrimPrefix(r.URL.Path, base+"/"), "/")
			if r.URL.Path == base {
				parts = nil
			}
			s := options.Service
			if r.Method == "POST" && len(parts) == 1 && parts[0] == "uploads" {
				media, _, e := mime.ParseMediaType(r.Header.Get("Content-Type"))
				if e != nil || media != "application/json" {
					return fail("INVALID_INPUT")
				}
				decoder := json.NewDecoder(http.MaxBytesReader(w, r.Body, 16384))
				decoder.DisallowUnknownFields()
				var input UploadInput
				if e = decoder.Decode(&input); e != nil {
					return fail("INVALID_INPUT")
				}
				var extra any
				if decoder.Decode(&extra) != io.EOF {
					return fail("INVALID_INPUT")
				}
				v, e := s.CreateUpload(ctx, actor, input)
				if e != nil {
					return e
				}
				send(v, 201)
				return nil
			}
			if len(parts) >= 2 && parts[0] == "uploads" {
				id := parts[1]
				if r.Method == "PUT" && len(parts) == 3 && parts[2] == "content" {
					info, e := s.Info(ctx, actor, id)
					if e != nil {
						return e
					}
					if r.Header.Get("Content-Type") != info.ContentType || (r.ContentLength >= 0 && r.ContentLength != info.Size) {
						return fail("INVALID_INPUT")
					}
					if e = s.Upload(ctx, actor, id, http.MaxBytesReader(w, r.Body, info.Size+1)); e != nil {
						return e
					}
					send(nil, 204)
					return nil
				}
				if r.Method == "POST" && len(parts) == 3 {
					var info FileInfo
					var e error
					switch parts[2] {
					case "complete":
						info, e = s.Complete(ctx, actor, id)
					case "recover":
						info, e = s.Recover(ctx, actor, id)
					default:
						return fail("NOT_FOUND")
					}
					if e != nil {
						return e
					}
					send(info, 200)
					return nil
				}
				if r.Method == "DELETE" && len(parts) == 2 {
					if e = s.Cancel(ctx, actor, id); e != nil {
						return e
					}
					send(nil, 204)
					return nil
				}
				return fail("NOT_FOUND")
			}
			if r.Method == "GET" && len(parts) == 0 {
				n := 0
				if r.URL.Query().Has("limit") {
					n, e = strconv.Atoi(r.URL.Query().Get("limit"))
					if e != nil || n < 1 {
						return fail("INVALID_INPUT")
					}
				}
				v, e := s.List(ctx, actor, r.URL.Query().Get("cursor"), n)
				if e != nil {
					return e
				}
				send(v, 200)
				return nil
			}
			if r.Method == "GET" && len(parts) == 1 {
				v, e := s.Info(ctx, actor, parts[0])
				if e != nil {
					return e
				}
				send(v, 200)
				return nil
			}
			if r.Method == "GET" && len(parts) == 2 && parts[1] == "download" {
				v, e := s.DownloadLocation(ctx, actor, parts[0])
				if e != nil {
					return e
				}
				send(v, 200)
				return nil
			}
			if r.Method == "GET" && len(parts) == 2 && parts[1] == "content" {
				info, e := s.Info(ctx, actor, parts[0])
				if e != nil {
					return e
				}
				v, e := s.Download(ctx, actor, parts[0])
				if e != nil {
					return e
				}
				defer v.Body.Close()
				w.Header().Set("Content-Type", v.Info.ContentType)
				w.Header().Set("Content-Length", strconv.FormatInt(v.Info.Size, 10))
				w.Header().Set("Content-Disposition", "attachment; filename*=UTF-8''"+strings.ReplaceAll(url.QueryEscape(info.Name), "+", "%20"))
				if _, e = io.Copy(w, v.Body); e != nil {
					panic(http.ErrAbortHandler)
				}
				return nil
			}
			if r.Method == "DELETE" && len(parts) == 1 {
				if e = s.Delete(ctx, actor, parts[0]); e != nil {
					return e
				}
				send(nil, 204)
				return nil
			}
			return fail("NOT_FOUND")
		}
		if e := operation(); e != nil {
			var typed *Error
			normalized := Normalize(e)
			if !errors.As(normalized, &typed) {
				typed = &Error{Code: "UNAVAILABLE"}
			}
			status := map[string]int{"INVALID_INPUT": 400, "NOT_FOUND": 404, "FORBIDDEN": 403, "CONFLICT": 409, "EXPIRED": 410, "UNSUPPORTED": 501, "UNAVAILABLE": 503, "ABORTED": 499, "CONFIGURATION": 503}[typed.Code]
			if status == 0 {
				status = 503
			}
			send(map[string]string{"code": typed.Code, "requestId": requestID}, status)
		}
	}), nil
}

File: architecture/modules/storage/node/src/http.ts
import type {IncomingMessage,ServerResponse} from 'node:http';
import {pipeline} from 'node:stream/promises';
import {randomUUID} from 'node:crypto';
import {StorageError,sdkError,pageLimit} from './contracts.ts';
import {FileService} from './service.ts';
export interface FileHTTPOptions {service:FileService;authenticate:(request:IncomingMessage)=>Promise<{id:string}|undefined>;basePath?:string;trustedOrigins?:readonly string[];allowCredentials?:boolean}
export function createFileHandler(options:FileHTTPOptions){
  const base=options.basePath||'/files';if(!/^\/[a-z0-9/-]+$/.test(base)||base.endsWith('/'))throw new StorageError('CONFIGURATION');
  return async function handle(request:IncomingMessage,response:ServerResponse):Promise<boolean>{
    const url=new URL(request.url||'/','http://localhost');if(url.pathname!==base&&!url.pathname.startsWith(base+'/'))return false;
    const requestId=randomUUID();response.setHeader('x-request-id',requestId);response.setHeader('cache-control','no-store');response.setHeader('x-content-type-options','nosniff');
    const send=(value:unknown,status=200)=>{response.statusCode=status;if(status===204){response.end();return;}response.setHeader('content-type','application/json');response.end(JSON.stringify(value));};
    const controller=new AbortController();request.on('aborted',()=>controller.abort());response.on('close',()=>{if(!response.writableFinished)controller.abort();});
    try{
      const origin=request.headers.origin;
      if(origin){if(!options.trustedOrigins?.includes(origin))throw new StorageError('FORBIDDEN');response.setHeader('access-control-allow-origin',origin);response.setHeader('vary','Origin');if(options.allowCredentials)response.setHeader('access-control-allow-credentials','true');}
      if(request.method==='OPTIONS'){response.setHeader('access-control-allow-methods','GET,POST,PUT,DELETE,OPTIONS');response.setHeader('access-control-allow-headers','authorization,content-type');send(undefined,204);return true;}
      const actor=await options.authenticate(request);if(!actor?.id)throw new StorageError('FORBIDDEN');
      const parts=url.pathname.slice(base.length).split('/').filter(Boolean).map(p=>decodeURIComponent(p)),method=request.method;
      if(method==='POST'&&parts.length===1&&parts[0]==='uploads'){
        if(request.headers['content-type']?.split(';')[0]!=='application/json')throw new StorageError('INVALID_INPUT');
        const chunks:Buffer[]=[];let size=0;for await(const chunk of request){size+=chunk.length;if(size>16384)throw new StorageError('INVALID_INPUT');chunks.push(Buffer.from(chunk));}
        let input;try{input=JSON.parse(Buffer.concat(chunks).toString());}catch{throw new StorageError('INVALID_INPUT');}
        if(Object.keys(input||{}).some(k=>!['name','size','contentType','storeId'].includes(k)))throw new StorageError('INVALID_INPUT');
        send(await options.service.createUpload(actor.id,input),201);
      }else if(parts[0]==='uploads'&&parts.length>=2){
        const id=parts[1];
        if(method==='PUT'&&parts.length===3&&parts[2]==='content'){
          const info=await options.service.info(actor.id,id);
          if(request.headers['content-type']!==info.contentType||(request.headers['content-length']!==undefined&&Number(request.headers['content-length'])!==info.size))throw new StorageError('INVALID_INPUT');
          await options.service.upload(actor.id,id,request,controller.signal);send(undefined,204);
        }else if(method==='POST'&&parts.length===3&&parts[2]==='complete')send(await options.service.complete(actor.id,id,controller.signal));
        else if(method==='POST'&&parts.length===3&&parts[2]==='recover')send(await options.service.recover(actor.id,id));
        else if(method==='DELETE'&&parts.length===2){await options.service.cancel(actor.id,id);send(undefined,204);}
        else throw new StorageError('NOT_FOUND');
      }else if(method==='GET'&&parts.length===0)send(await options.service.list(actor.id,{cursor:url.searchParams.get('cursor')||undefined,limit:pageLimit(url.searchParams.has('limit')?Number(url.searchParams.get('limit')):undefined)}));
      else if(method==='GET'&&parts.length===1)send(await options.service.info(actor.id,parts[0]));
      else if(method==='GET'&&parts.length===2&&parts[1]==='download')send(await options.service.downloadLocation(actor.id,parts[0]));
      else if(method==='GET'&&parts.length===2&&parts[1]==='content'){
        const info=await options.service.info(actor.id,parts[0]),download=await options.service.download(actor.id,parts[0],controller.signal);
        response.setHeader('content-type',download.info.contentType);response.setHeader('content-length',download.info.size);response.setHeader('content-disposition',"attachment; filename*=UTF-8''"+encodeURIComponent(info.name));await pipeline(download.body,response,{signal:controller.signal});
      }else if(method==='DELETE'&&parts.length===1){await options.service.delete(actor.id,parts[0]);send(undefined,204);}
      else throw new StorageError('NOT_FOUND');
    }catch(error){
      const e=error instanceof URIError?new StorageError('INVALID_INPUT'):sdkError(error);
      if(response.headersSent){response.destroy();return true;}
      const status={INVALID_INPUT:400,NOT_FOUND:404,FORBIDDEN:403,CONFLICT:409,EXPIRED:410,UNSUPPORTED:501,UNAVAILABLE:503,ABORTED:499,CONFIGURATION:503}[e.code];send({code:e.code,requestId},status);
    }return true;
  };
}
